// @smartcart/api/src/config/pino.config.ts
import type { Params } from 'nestjs-pino';
import { randomUUID } from 'node:crypto';
import type { IncomingMessage, ServerResponse } from 'node:http';

/**
 * Structured logging config (README §2.6 Logging, §2.5 A02/A09 PII redaction).
 * Every sensitive field is redacted to `[Redacted]` before a line is written.
 * Each request gets a `correlationId` propagated via the `x-correlation-id` header.
 */

// Exact redaction set required by README §2.5 A02.
export const REDACTED_FIELDS = [
  'password',
  'password_hash',
  'passwordHash',
  'refreshToken',
  'accessToken',
  'email',
  'phone',
  'pushToken',
];

// Build wildcard paths so the fields are redacted wherever they appear in the log object.
const redactPaths = REDACTED_FIELDS.flatMap((f) => [
  f,
  `*.${f}`,
  `req.body.${f}`,
  `req.headers.${f}`,
]);

export const pinoConfig: Params = {
  pinoHttp: {
    level: process.env.NODE_ENV === 'production' ? 'info' : 'debug',
    // Pretty logs in dev; raw JSON (for Loki/Promtail) in prod.
    transport:
      process.env.NODE_ENV === 'production'
        ? undefined
        : { target: 'pino-pretty', options: { singleLine: true } },
    redact: { paths: redactPaths, censor: '[Redacted]' },
    genReqId: (req: IncomingMessage, res: ServerResponse) => {
      const existing =
        (req.headers['x-correlation-id'] as string | undefined) ?? randomUUID();
      res.setHeader('x-correlation-id', existing);
      return existing;
    },
    customProps: (req: IncomingMessage) => ({
      correlationId: (req as IncomingMessage & { id?: string }).id,
      userId: (req as { user?: { sub?: string } }).user?.sub,
      role: (req as { user?: { role?: string } }).user?.role,
    }),
  },
};
