// @smartcart/api/src/common/middleware/rate-limiter.middleware.ts
import {
  HttpException,
  HttpStatus,
  Injectable,
  NestMiddleware,
} from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import type { NextFunction, Request, Response } from 'express';
import { RedisService } from '../redis/redis.service';

/**
 * Redis-based fixed-window rate limiter (README §2.5 Rate Limiting, A07).
 * Global default 100 req/min per user/IP; stricter on auth routes (10 req/min)
 * to throttle brute-force. Emits `X-RateLimit-*` headers; 429 over the cap.
 */
@Injectable()
export class RateLimiterMiddleware implements NestMiddleware {
  private readonly max: number;
  private readonly authMax: number;
  private readonly windowMs: number;

  constructor(
    private readonly redis: RedisService,
    config: ConfigService,
  ) {
    this.max = config.get<number>('RATE_LIMIT_MAX', 100);
    this.authMax = config.get<number>('AUTH_RATE_LIMIT_MAX', 10);
    this.windowMs = config.get<number>('RATE_LIMIT_WINDOW_MS', 60_000);
  }

  async use(req: Request, res: Response, next: NextFunction): Promise<void> {
    const isAuth = req.path.includes('/auth/');
    const limit = isAuth ? this.authMax : this.max;

    // Middleware runs before guards, so fall back to IP when no user is known yet.
    const subject =
      (req as Request & { user?: { sub?: string } }).user?.sub ?? req.ip;
    const key = `ratelimit:${isAuth ? 'auth' : 'global'}:${subject}`;

    const count = await this.redis.incr(key);
    if (count === 1) {
      await this.redis.pexpire(key, this.windowMs);
    }

    const remaining = Math.max(0, limit - count);
    res.setHeader('X-RateLimit-Limit', String(limit));
    res.setHeader('X-RateLimit-Remaining', String(remaining));

    if (count > limit) {
      throw new HttpException(
        'Rate limit exceeded. Slow down.',
        HttpStatus.TOO_MANY_REQUESTS,
      );
    }
    next();
  }
}
