// @smartcart/api/src/modules/checkout/presentation/gateways/session.gateway.ts
import { Inject, Logger } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import {
  ConnectedSocket,
  MessageBody,
  OnGatewayConnection,
  SubscribeMessage,
  WebSocketGateway,
  WebSocketServer,
} from '@nestjs/websockets';
import * as jwt from 'jsonwebtoken';
import { Server, Socket } from 'socket.io';
import {
  ISessionNotifier,
  SessionStatusChange,
} from '../../application/interfaces/session-notifier.interface';
import {
  ISessionRepository,
  SESSION_REPOSITORY,
} from '../../application/interfaces/session-repository.interface';

const room = (sessionId: string) => `session:${sessionId}`;

/**
 * Real-time checkout status gateway (README §2.8 Workflow 1 step 5).
 * - JWT handshake auth: an unauthenticated socket is dropped on connect.
 * - Ownership-checked subscription: a client may only join the room for a
 *   session it owns, so no one can observe another user's checkout.
 * Implements `ISessionNotifier` so the application can push without importing it.
 */
@WebSocketGateway({ cors: { origin: '*' } })
export class SessionGateway
  implements OnGatewayConnection, ISessionNotifier
{
  @WebSocketServer() server!: Server;
  private readonly logger = new Logger(SessionGateway.name);
  private readonly accessSecret: string;

  constructor(
    config: ConfigService,
    @Inject(SESSION_REPOSITORY) private readonly sessions: ISessionRepository,
  ) {
    this.accessSecret = config.getOrThrow<string>('JWT_ACCESS_SECRET');
  }

  /** Authenticate the JWT on the handshake; reject unauthenticated sockets. */
  handleConnection(client: Socket): void {
    try {
      const raw =
        (client.handshake.auth?.token as string | undefined) ??
        (client.handshake.headers.authorization as string | undefined)?.replace(
          /^Bearer\s+/i,
          '',
        );
      if (!raw) throw new Error('missing token');
      const payload = jwt.verify(raw, this.accessSecret) as { sub: string };
      client.data.userId = payload.sub;
    } catch {
      client.disconnect(true);
    }
  }

  /** Join the per-session room only after verifying ownership. */
  @SubscribeMessage('subscribe:session')
  async onSubscribe(
    @ConnectedSocket() client: Socket,
    @MessageBody() data: { sessionId: string },
  ): Promise<{ subscribed: string } | { error: string }> {
    const userId = client.data.userId as string | undefined;
    if (!userId) {
      client.disconnect(true);
      return { error: 'unauthenticated' };
    }
    const session = await this.sessions.findById(data.sessionId);
    if (!session || session.userId !== userId) {
      return { error: 'forbidden' };
    }
    await client.join(room(data.sessionId));
    return { subscribed: data.sessionId };
  }

  /** ISessionNotifier — emit to the per-session room (README W1.5). */
  notifyStatusChanged(change: SessionStatusChange): void {
    this.server.to(room(change.sessionId)).emit('sessionStatusChanged', {
      status: change.status,
      pointsAwarded: change.pointsAwarded,
    });
  }
}
