// @smartcart/api/eslint.config.mjs
// ESLint flat config — module-boundary + layer enforcement (README §2.2).
// Boundaries are compile-time, not runtime: a cross-layer import fails the build.
import tseslint from 'typescript-eslint';

export default tseslint.config(
  {
    ignores: ['dist/**', 'node_modules/**', 'coverage/**', 'prisma/**'],
  },
  {
    files: ['src/**/*.ts'],
    languageOptions: {
      parser: tseslint.parser,
      parserOptions: { sourceType: 'module' },
    },
    plugins: { '@typescript-eslint': tseslint.plugin },
    rules: {
      '@typescript-eslint/no-explicit-any': 'warn',
      '@typescript-eslint/no-unused-vars': [
        'error',
        { argsIgnorePattern: '^_' },
      ],
    },
  },
  // Rule 1 — Domain layer: zero framework / infrastructure imports.
  {
    files: ['src/modules/*/domain/**/*.ts'],
    rules: {
      'no-restricted-imports': [
        'error',
        {
          patterns: [
            {
              group: [
                '@nestjs/*',
                '@prisma/client',
                'ioredis',
                'bullmq',
                'axios',
                '**/infrastructure/**',
                '**/application/**',
                '**/presentation/**',
              ],
              message:
                'Domain must be pure TypeScript — no NestJS, Prisma, or other layers (README §2.2 Rule 1).',
            },
          ],
        },
      ],
    },
  },
  // Rule 2 — Application layer: interfaces only, never concrete infrastructure.
  {
    files: ['src/modules/*/application/**/*.ts'],
    rules: {
      'no-restricted-imports': [
        'error',
        {
          patterns: [
            {
              group: ['**/infrastructure/**', '@prisma/client'],
              message:
                'Application depends on interfaces, never concrete infrastructure classes or PrismaClient (README §2.2 Rule 2).',
            },
            {
              group: ['../../*/domain/**', '../../*/infrastructure/**'],
              message:
                "Do not import another module's internals (README §2.2 Module Boundary).",
            },
          ],
        },
      ],
    },
  },
  // Rule 4 — Presentation layer: no domain entities / repositories / Prisma.
  {
    files: ['src/modules/*/presentation/**/*.ts'],
    rules: {
      'no-restricted-imports': [
        'error',
        {
          patterns: [
            {
              group: ['**/domain/entities/**', '@prisma/client', 'ioredis'],
              message:
                'Controllers delegate to application services — no domain entities or Prisma (README §2.2 Rule 4).',
            },
          ],
        },
      ],
    },
  },
);
