// @smartcart/api/src/common/filters/global-exception.filter.ts
import {
  ArgumentsHost,
  Catch,
  ExceptionFilter,
  HttpException,
  HttpStatus,
  Logger,
} from '@nestjs/common';
import * as Sentry from '@sentry/node';
import type { Request, Response } from 'express';
import { DomainError } from '../../modules/checkout/domain/errors/checkout.errors';

/**
 * Maps a thrown DomainError `code` to its HTTP status / error code
 * (README §2.8 Workflow 1 Error Handling Matrix).
 */
const DOMAIN_STATUS: Record<string, number> = {
  QR_TOKEN_EXPIRED: HttpStatus.UNPROCESSABLE_ENTITY, // 422
  INVALID_QR_TOKEN: HttpStatus.UNAUTHORIZED, // 401
  QR_ITEM_MISMATCH: HttpStatus.UNPROCESSABLE_ENTITY, // 422
  INVALID_TRANSITION: HttpStatus.CONFLICT, // 409
  EMPTY_CART: HttpStatus.BAD_REQUEST, // 400
  ITEM_NOT_FOUND: HttpStatus.NOT_FOUND, // 404
  UNKNOWN_STRATEGY: HttpStatus.INTERNAL_SERVER_ERROR, // 500
};

/**
 * Single error pipeline (README §2.5 A05/A09, §2.6 Error Tracking).
 * - Strips stack traces from 500s in production.
 * - Always returns `{ errorCode, message, correlationId }` — never an internal stack.
 * - Reports only 5xx (ERROR severity) to Sentry; expected 4xx are excluded.
 */
@Catch()
export class GlobalExceptionFilter implements ExceptionFilter {
  private readonly logger = new Logger(GlobalExceptionFilter.name);
  private readonly isProduction = process.env.NODE_ENV === 'production';

  catch(exception: unknown, host: ArgumentsHost): void {
    const ctx = host.switchToHttp();
    const response = ctx.getResponse<Response>();
    const request = ctx.getRequest<Request>();
    const correlationId =
      (request.headers['x-correlation-id'] as string | undefined) ??
      (request as Request & { id?: string }).id;

    const { status, errorCode, message } = this.resolve(exception);

    if (status >= 500) {
      // ERROR severity → Sentry + server log with stack (server-side only).
      Sentry.captureException(exception, { tags: { correlationId } });
      this.logger.error(
        { event: 'UNHANDLED', errorCode, correlationId },
        exception instanceof Error ? exception.stack : String(exception),
      );
    }

    response.status(status).json({
      errorCode,
      message: this.isProduction && status >= 500 ? 'Internal server error' : message,
      correlationId,
    });
  }

  private resolve(exception: unknown): {
    status: number;
    errorCode: string;
    message: string;
  } {
    if (exception instanceof DomainError) {
      const status =
        DOMAIN_STATUS[exception.code] ?? HttpStatus.UNPROCESSABLE_ENTITY;
      return { status, errorCode: exception.code, message: exception.message };
    }

    if (exception instanceof HttpException) {
      const status = exception.getStatus();
      const res = exception.getResponse();
      const message =
        typeof res === 'string'
          ? res
          : ((res as { message?: string | string[] }).message?.toString() ??
            exception.message);
      const errorCode =
        (typeof res === 'object' &&
          (res as { errorCode?: string }).errorCode) ||
        HttpStatus[status] ||
        'ERROR';
      return { status, errorCode, message };
    }

    return {
      status: HttpStatus.INTERNAL_SERVER_ERROR,
      errorCode: 'INTERNAL_ERROR',
      message:
        exception instanceof Error ? exception.message : 'Unexpected error',
    };
  }
}
