// @smartcart/api/src/modules/checkout/domain/factories/qr-ticket.factory.ts
import { ShoppingSession } from '../entities/shopping-session.entity';
import { QrTicket } from '../value-objects/qr-ticket.vo';

/**
 * Claims embedded in the signed QR JWT (README §2.3 QR generation).
 */
export interface QrSignPayload {
  sessionId: string;
  itemHash: string;
  userId: string;
}

/**
 * Result of signing — the token plus the expiry decoded from its `exp` claim.
 * The factory reads expiry from here and never computes its own, so 10-minute
 * expiry has exactly one source of truth: the signer (README §2.3 QR rules).
 */
export interface SignedQrToken {
  token: string;
  expiresAt: Date;
}

/**
 * Structural signer port. The infrastructure `JwtQrSigner` (and the application
 * `IQrSigner` interface) conform to this shape; declaring it here keeps the
 * factory in the pure domain layer with no upward import (README §2.2 Rule 1).
 */
export interface QrTokenSigner {
  sign(payload: QrSignPayload): SignedQrToken;
}

/**
 * QrTicketFactory — Factory pattern (README §2.3 QR / §2.8 Workflow 4 step 5).
 * Transitions the session to PENDING_CHECKOUT (freezing & hashing the cart),
 * signs the JWT, and returns the QrTicket value object.
 */
export class QrTicketFactory {
  static create(session: ShoppingSession, signer: QrTokenSigner): QrTicket {
    // Freeze the cart and compute its hash (throws if empty / wrong state).
    session.requestCheckout();
    const itemHash = session.itemHash;
    if (itemHash === null) {
      // requestCheckout always sets the hash; this guards the invariant.
      throw new Error('Item hash was not set after requestCheckout().');
    }

    const { token, expiresAt } = signer.sign({
      sessionId: session.id,
      itemHash,
      userId: session.userId,
    });

    return new QrTicket({ token, expiresAt, sessionId: session.id, itemHash });
  }
}
