// @smartcart/api/src/modules/checkout/domain/entities/shopping-session.entity.ts
import { createHash } from 'node:crypto';
import type { SessionStatus } from '@smartcart/shared-types';
import {
  EmptyCartError,
  ItemNotFoundError,
  QrItemMismatchError,
} from '../errors/checkout.errors';
import {
  SessionEvent,
  SessionStateMachine,
} from '../state-machine/session-state-machine';

/**
 * SessionItem — entity composed under the ShoppingSession aggregate (README §2.4).
 * `unitPrice` is NOT persisted by the README schema; it is carried in memory from
 * the product at scan time so SPEND_MULTIPLIER can price the line (README §2.3).
 */
export class SessionItem {
  constructor(
    readonly id: string,
    readonly productId: string,
    readonly barcode: string,
    readonly quantity: number,
    public pointsValue: number,
    readonly unitPrice: number = 0,
  ) {}
}

export interface AddItemInput {
  id: string;
  productId: string;
  barcode: string;
  quantity: number;
  pointsValue: number;
  unitPrice?: number;
}

export interface ShoppingSessionProps {
  id: string;
  userId: string;
  storeId: string;
  status: SessionStatus;
  items: SessionItem[];
  itemHash: string | null;
  createdAt: Date;
  updatedAt: Date;
}

/**
 * ShoppingSession — aggregate root (README §2.2 Rule 1, §2.3 State Machine).
 * The most important domain object: it owns the cart, enforces the state machine
 * guards, and computes the anti-tamper item hash. No NestJS / Prisma imports.
 */
export class ShoppingSession {
  readonly id: string;
  readonly userId: string;
  readonly storeId: string;

  private _status: SessionStatus;
  private _items: SessionItem[];
  private _itemHash: string | null;
  readonly createdAt: Date;
  private _updatedAt: Date;

  private constructor(props: ShoppingSessionProps) {
    this.id = props.id;
    this.userId = props.userId;
    this.storeId = props.storeId;
    this._status = props.status;
    this._items = props.items;
    this._itemHash = props.itemHash;
    this.createdAt = props.createdAt;
    this._updatedAt = props.updatedAt;
  }

  // ── Factory methods ────────────────────────────────────────────────────────

  /** Create a brand-new ACTIVE session (invariant: empty cart, no hash). */
  static create(input: {
    id: string;
    userId: string;
    storeId: string;
    now?: Date;
  }): ShoppingSession {
    const now = input.now ?? new Date();
    return new ShoppingSession({
      id: input.id,
      userId: input.userId,
      storeId: input.storeId,
      status: 'ACTIVE',
      items: [],
      itemHash: null,
      createdAt: now,
      updatedAt: now,
    });
  }

  /** Rehydrate from persistence (used by the SessionMapper, README §2.2 Rule 3). */
  static rehydrate(props: ShoppingSessionProps): ShoppingSession {
    return new ShoppingSession({ ...props, items: [...props.items] });
  }

  // ── Accessors ──────────────────────────────────────────────────────────────

  get status(): SessionStatus {
    return this._status;
  }
  /** Defensive copy — callers cannot mutate the cart directly. */
  get items(): readonly SessionItem[] {
    return [...this._items];
  }
  get itemHash(): string | null {
    return this._itemHash;
  }
  get updatedAt(): Date {
    return this._updatedAt;
  }

  // ── Behavior (state-machine guarded) ────────────────────────────────────────

  /** Add a scanned line. Allowed only while ACTIVE (README §2.8 Workflow 4). */
  addItem(input: AddItemInput): SessionItem {
    this._status = SessionStateMachine.transition(
      this._status,
      SessionEvent.ADD_ITEM,
    ); // ACTIVE → ACTIVE, throws otherwise
    const item = new SessionItem(
      input.id,
      input.productId,
      input.barcode,
      input.quantity,
      input.pointsValue,
      input.unitPrice ?? 0,
    );
    this._items.push(item);
    this.touch();
    return item;
  }

  /** Remove a line by id. Allowed only while ACTIVE. */
  removeItem(itemId: string): void {
    SessionStateMachine.transition(this._status, SessionEvent.ADD_ITEM); // assert ACTIVE
    const idx = this._items.findIndex((i) => i.id === itemId);
    if (idx === -1) throw new ItemNotFoundError(itemId);
    this._items.splice(idx, 1);
    this.touch();
  }

  /**
   * Freeze the cart for checkout: ACTIVE → PENDING_CHECKOUT, compute & store the
   * item hash. Requires ≥ 1 item (README §2.3 guard). Idempotent hashing.
   */
  requestCheckout(): void {
    if (this._items.length === 0) throw new EmptyCartError();
    this._status = SessionStateMachine.transition(
      this._status,
      SessionEvent.REQUEST_CHECKOUT,
    );
    this._itemHash = this.computeItemHash();
    this.touch();
  }

  /** PENDING_CHECKOUT → COMPLETED (after a successful POS validation). */
  completeValidation(): void {
    this._status = SessionStateMachine.transition(
      this._status,
      SessionEvent.COMPLETE_VALIDATION,
    );
    this.touch();
  }

  /** PENDING_CHECKOUT → VALIDATION_FAILED (terminal; hash mismatch). */
  markValidationFailed(): void {
    this._status = SessionStateMachine.transition(
      this._status,
      SessionEvent.MARK_VALIDATION_FAILED,
    );
    this.touch();
  }

  /** Expire via cron. Idempotent for COMPLETED (README §2.3 table). */
  expire(): void {
    const next = SessionStateMachine.transition(
      this._status,
      SessionEvent.EXPIRE,
    );
    if (next === this._status) return; // COMPLETED stays COMPLETED — no touch
    this._status = next;
    this.touch();
  }

  // ── Anti-tamper hashing (README §2.3 QR) ────────────────────────────────────

  /**
   * Deterministic cart hash: sort barcodes alphabetically, join with "|",
   * SHA-256 hex. The POS side runs the EXACT same algorithm on its scanned
   * barcodes — any add/remove/swap changes the hash. Pure, no side effects.
   */
  computeItemHash(): string {
    return ShoppingSession.hashBarcodes(this._items.map((i) => i.barcode));
  }

  /** Shared hash function so domain + POS path cannot diverge on sort/separator. */
  static hashBarcodes(barcodes: string[]): string {
    const canonical = [...barcodes].sort().join('|');
    return createHash('sha256').update(canonical).digest('hex');
  }

  /**
   * Compare the POS-scanned barcodes against the frozen hash (README §2.8 W1.3).
   * Throws QrItemMismatchError on any difference.
   */
  validateItems(scannedBarcodes: string[]): void {
    if (this._itemHash === null) {
      throw new QrItemMismatchError();
    }
    const scannedHash = ShoppingSession.hashBarcodes(scannedBarcodes);
    if (scannedHash !== this._itemHash) {
      throw new QrItemMismatchError();
    }
  }

  /** Total pending points across the cart (display + provisional accrual). */
  totalPoints(): number {
    return this._items.reduce((sum, i) => sum + i.pointsValue, 0);
  }

  private touch(): void {
    this._updatedAt = new Date();
  }
}
