// @smartcart/api/src/modules/auth/infrastructure/crypto/jwt.service.ts
import { randomUUID } from 'node:crypto';
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import type { Role } from '@smartcart/shared-types';
import * as jwt from 'jsonwebtoken';

export interface AccessClaims {
  sub: string;
  role: Role;
}

export interface RefreshClaims {
  sub: string;
  jti: string;
}

export interface SignedRefresh {
  token: string;
  jti: string;
}

/**
 * JWT issuing/verification (README §2.5 Authentication, §2.8 Workflow 3 step 3).
 * Access token: 15 min, embeds `sub` + `role`. Refresh token: 7 days, embeds a
 * unique `jti` so a single token can be revoked/rotated. Secrets (≥ 32 chars)
 * are validated at startup by `env.validation.ts`.
 */
@Injectable()
export class JwtTokenService {
  private readonly accessSecret: string;
  private readonly refreshSecret: string;
  private readonly accessTtl: string;
  private readonly refreshTtl: string;

  constructor(config: ConfigService) {
    this.accessSecret = config.getOrThrow<string>('JWT_ACCESS_SECRET');
    this.refreshSecret = config.getOrThrow<string>('JWT_REFRESH_SECRET');
    this.accessTtl = config.get<string>('JWT_ACCESS_TTL', '15m');
    this.refreshTtl = config.get<string>('JWT_REFRESH_TTL', '7d');
  }

  signAccessToken(userId: string, role: Role): string {
    return jwt.sign({ sub: userId, role }, this.accessSecret, {
      expiresIn: this.accessTtl,
    } as jwt.SignOptions);
  }

  signRefreshToken(userId: string): SignedRefresh {
    const jti = randomUUID();
    const token = jwt.sign({ sub: userId, jti }, this.refreshSecret, {
      expiresIn: this.refreshTtl,
    } as jwt.SignOptions);
    return { token, jti };
  }

  verifyAccessToken(token: string): AccessClaims {
    return jwt.verify(token, this.accessSecret) as AccessClaims;
  }

  verifyRefreshToken(token: string): RefreshClaims {
    return jwt.verify(token, this.refreshSecret) as RefreshClaims;
  }
}
