// @smartcart/api/src/modules/auth/application/services/auth.service.ts
import { createHash } from 'node:crypto';
import {
  ConflictException,
  HttpException,
  HttpStatus,
  Inject,
  Injectable,
  Logger,
  UnauthorizedException,
} from '@nestjs/common';
import type {
  AuthTokens,
  AuthUser,
  LoginRequest,
  RegisterRequest,
} from '@smartcart/shared-types';
import { RedisService } from '../../../../common/redis/redis.service';
import { JwtTokenService } from '../../infrastructure/crypto/jwt.service';
import { PasswordService } from '../../infrastructure/crypto/password.service';
import {
  IUserRepository,
  USER_REPOSITORY,
} from '../interfaces/user-repository.interface';

export interface AuthResult {
  user: AuthUser;
  tokens: AuthTokens;
}

// Account lockout policy (README §2.5 A07): 5 fails / 15 min → 30-min lockout.
const MAX_FAILED = 5;
const FAIL_WINDOW_SECONDS = 15 * 60;
const LOCKOUT_SECONDS = 30 * 60;
// Refresh-token revocation window (README §2.5: 7 days).
const REFRESH_TTL_SECONDS = 7 * 24 * 60 * 60;

const sha256 = (s: string) => createHash('sha256').update(s).digest('hex');
const lockoutKey = (email: string) => `auth:lockout:${email.toLowerCase()}`;
const refreshKey = (userId: string, jti: string) => `refresh:${userId}:${jti}`;

/**
 * Authentication service (README §2.5 Authentication / A07, §2.8 Workflow 3).
 * The mobile app only ever issues USER-scoped tokens — privileged roles are
 * never minted here.
 */
@Injectable()
export class AuthService {
  private readonly logger = new Logger(AuthService.name);

  constructor(
    @Inject(USER_REPOSITORY) private readonly users: IUserRepository,
    private readonly passwords: PasswordService,
    private readonly jwt: JwtTokenService,
    private readonly redis: RedisService,
  ) {}

  /** Register a USER account; duplicate email → 409 (README §2.8 W3.1). */
  async register(dto: RegisterRequest): Promise<AuthResult> {
    const existing = await this.users.findByEmail(dto.email);
    if (existing) {
      throw new ConflictException('Email already registered.');
    }
    const passwordHash = await this.passwords.hash(dto.password);
    const user = await this.users.create({
      email: dto.email,
      fullName: dto.fullName,
      passwordHash,
      role: 'USER', // never a privileged role from the mobile surface
    });
    return this.buildResult(user.id, user.email, 'USER');
  }

  /**
   * Login. Unknown email and wrong password return the SAME error (no user
   * enumeration). Lockout after 5 fails in 15 min (README §2.5 A07).
   */
  async login(dto: LoginRequest): Promise<AuthResult> {
    await this.assertNotLockedOut(dto.email);

    const user = await this.users.findByEmail(dto.email);
    const ok =
      user !== null &&
      (await this.passwords.verify(dto.password, user.passwordHash));

    if (!user || !ok) {
      await this.registerFailedAttempt(dto.email);
      throw new UnauthorizedException('Invalid credentials.');
    }

    await this.redis.del(lockoutKey(dto.email));
    return this.buildResult(user.id, user.email, user.role);
  }

  /**
   * Rotate the refresh token (README §2.5 A07, §2.8 W3.4). Old token is deleted
   * before the new one is issued; replaying the old token → 401.
   */
  async refresh(refreshToken: string): Promise<AuthResult> {
    let claims: { sub: string; jti: string };
    try {
      claims = this.jwt.verifyRefreshToken(refreshToken);
    } catch {
      throw new UnauthorizedException('Invalid refresh token.');
    }

    const key = refreshKey(claims.sub, claims.jti);
    const storedHash = await this.redis.get(key);
    if (!storedHash || storedHash !== sha256(refreshToken)) {
      // Absent/mismatched ⇒ revoked or already rotated (replay).
      throw new UnauthorizedException('Refresh token has been revoked.');
    }
    await this.redis.del(key); // single-use: invalidate before re-issuing

    const user = await this.users.findById(claims.sub);
    if (!user) throw new UnauthorizedException('Invalid refresh token.');
    return this.buildResult(user.id, user.email, user.role);
  }

  /** Logout = server-side revocation of the refresh token (README §2.8 W3.5). */
  async logout(userId: string, refreshToken?: string): Promise<void> {
    if (!refreshToken) return;
    try {
      const { jti, sub } = this.jwt.verifyRefreshToken(refreshToken);
      if (sub === userId) await this.redis.del(refreshKey(userId, jti));
    } catch {
      // Already-invalid token — nothing to revoke.
    }
  }

  // ── internals ───────────────────────────────────────────────────────────────

  private async buildResult(
    userId: string,
    email: string,
    role: AuthUser['role'],
  ): Promise<AuthResult> {
    const accessToken = this.jwt.signAccessToken(userId, role);
    const { token: refreshToken, jti } = this.jwt.signRefreshToken(userId);
    // Store only the HASH of the refresh token for revocation (README §2.5 A07).
    await this.redis.set(
      refreshKey(userId, jti),
      sha256(refreshToken),
      'EX',
      REFRESH_TTL_SECONDS,
    );
    return { user: { id: userId, email, role }, tokens: { accessToken, refreshToken } };
  }

  private async assertNotLockedOut(email: string): Promise<void> {
    const attempts = Number((await this.redis.get(lockoutKey(email))) ?? 0);
    if (attempts >= MAX_FAILED) {
      throw new HttpException(
        'Too many failed attempts. Try again later.',
        HttpStatus.TOO_MANY_REQUESTS,
      );
    }
  }

  private async registerFailedAttempt(email: string): Promise<void> {
    const key = lockoutKey(email);
    const count = await this.redis.incr(key);
    // First failure starts the 15-min window; on hitting the cap, extend to 30 min.
    if (count === 1) {
      await this.redis.expire(key, FAIL_WINDOW_SECONDS);
    } else if (count >= MAX_FAILED) {
      await this.redis.expire(key, LOCKOUT_SECONDS);
    }
  }
}
