// @smartcart/api/src/common/interceptors/audit.interceptor.ts
import {
  CallHandler,
  ExecutionContext,
  Injectable,
  Logger,
  NestInterceptor,
} from '@nestjs/common';
import type { Request } from 'express';
import { Observable } from 'rxjs';
import { tap } from 'rxjs/operators';

/**
 * Maps controller handler names → audit action labels for the sensitive
 * operations listed in README §2.5 Audit Logging.
 */
const SENSITIVE_ACTIONS: Record<string, string> = {
  register: 'register',
  login: 'login',
  logout: 'logout',
  refresh: 'refresh',
  validate: 'validateSession',
  redeem: 'redeemReward',
  update: 'updateProfile',
  remove: 'deleteAccount',
};

/**
 * Audit interceptor (README §2.5 A09). Emits a structured JSON entry for every
 * sensitive operation, with `userId`, `ip`, and `correlationId`. Registered
 * globally in `main.ts` so no new sensitive endpoint can be missed.
 */
@Injectable()
export class AuditInterceptor implements NestInterceptor {
  private readonly logger = new Logger('Audit');

  intercept(ctx: ExecutionContext, next: CallHandler): Observable<unknown> {
    const handlerName = ctx.getHandler().name;
    const action = SENSITIVE_ACTIONS[handlerName];

    if (!action) return next.handle();

    const request = ctx.switchToHttp().getRequest<
      Request & { user?: { sub?: string }; params: Record<string, string> }
    >();

    return next.handle().pipe(
      tap(() => {
        this.logger.log({
          event: 'AUDIT',
          action,
          userId: request.user?.sub,
          ip: request.ip,
          correlationId: request.headers['x-correlation-id'],
          sessionId: request.params?.id,
        });
      }),
    );
  }
}
