// @smartcart/api/src/common/guards/api-key.guard.ts
import { createHash } from 'node:crypto';
import {
  CanActivate,
  ExecutionContext,
  ForbiddenException,
  Injectable,
  SetMetadata,
  UnauthorizedException,
} from '@nestjs/common';
import { Reflector } from '@nestjs/core';
import type { Request } from 'express';
import { PrismaService } from '../prisma/prisma.service';

export type ApiKeyType = 'POS' | 'B2B';

/** Context attached to `request.apiKey` after a successful key check. */
export interface ApiKeyContext {
  type: ApiKeyType;
  storeId?: string;
}

/** Metadata key — restrict an endpoint to a specific API-key type. */
export const API_KEY_TYPE = 'apiKeyType';

/** `@RequireApiKey('POS' | 'B2B')` — narrow which key type may call the route. */
export const RequireApiKey = (type: ApiKeyType) =>
  SetMetadata(API_KEY_TYPE, type);

const sha256 = (s: string) => createHash('sha256').update(s).digest('hex');

/**
 * API-key authentication for POS / B2B surfaces (README §2.5 Authorization, A01).
 * Reads `X-API-Key`, SHA-256 hashes it, and looks it up in the `ApiKey` table.
 * Missing/invalid → 401; wrong type for the route → 403.
 */
@Injectable()
export class ApiKeyGuard implements CanActivate {
  constructor(
    private readonly prisma: PrismaService,
    private readonly reflector: Reflector,
  ) {}

  async canActivate(ctx: ExecutionContext): Promise<boolean> {
    const request = ctx
      .switchToHttp()
      .getRequest<Request & { apiKey?: ApiKeyContext }>();

    const provided = request.headers['x-api-key'];
    const key = Array.isArray(provided) ? provided[0] : provided;
    if (!key) throw new UnauthorizedException('Missing API key.');

    const record = await this.prisma.apiKey.findUnique({
      where: { hashedKey: sha256(key) },
    });
    if (!record) throw new UnauthorizedException('Invalid API key.');

    const required = this.reflector.getAllAndOverride<ApiKeyType | undefined>(
      API_KEY_TYPE,
      [ctx.getHandler(), ctx.getClass()],
    );
    if (required && record.type !== required) {
      throw new ForbiddenException(`This endpoint requires a ${required} key.`);
    }

    request.apiKey = {
      type: record.type as ApiKeyType,
      storeId: record.storeId ?? undefined,
    };
    return true;
  }
}
