// @smartcart/api/src/common/pipes/zod-validation.pipe.ts
import { BadRequestException, Injectable, PipeTransform } from '@nestjs/common';
import { ZodError, ZodSchema } from 'zod';

/**
 * Generic Zod validation pipe (README §2.2 Type Sharing, §2.5 Input Validation).
 * Enforces the shared `@smartcart/shared-types` schemas at the controller
 * boundary — strict types, length caps, UUIDs — rejecting injection / ReDoS
 * payloads with a 400 before they reach a service.
 */
@Injectable()
export class ZodValidationPipe<T = unknown> implements PipeTransform<unknown, T> {
  constructor(private readonly schema: ZodSchema<T>) {}

  transform(value: unknown): T {
    try {
      return this.schema.parse(value);
    } catch (err) {
      if (err instanceof ZodError) {
        throw new BadRequestException({
          errorCode: 'VALIDATION_FAILED',
          message: 'Invalid request payload.',
          issues: err.issues.map((i) => ({
            path: i.path.join('.'),
            message: i.message,
          })),
        });
      }
      throw err;
    }
  }
}
