// @smartcart/api/src/modules/checkout/infrastructure/mappers/session.mapper.ts
import type {
  Prisma,
  SessionItem as SessionItemRow,
  ShoppingSession as ShoppingSessionRow,
} from '@prisma/client';
import type { SessionStatus } from '@smartcart/shared-types';
import {
  SessionItem,
  ShoppingSession,
} from '../../domain/entities/shopping-session.entity';

type SessionRowWithItems = ShoppingSessionRow & { items: SessionItemRow[] };

/** Serializable shape stored in the Redis read-through cache. */
export interface SessionSnapshot {
  id: string;
  userId: string;
  storeId: string;
  status: SessionStatus;
  itemHash: string | null;
  createdAt: string;
  updatedAt: string;
  items: {
    id: string;
    productId: string;
    barcode: string;
    quantity: number;
    pointsValue: number;
  }[];
}

/**
 * Maps between Prisma rows and the ShoppingSession aggregate (README §2.2 Rule 3).
 * Prisma types never leak past this boundary — callers get domain entities.
 */
export class SessionMapper {
  /** DB row (+ items) → domain entity. */
  static toDomain(row: SessionRowWithItems): ShoppingSession {
    return ShoppingSession.rehydrate({
      id: row.id,
      userId: row.userId,
      storeId: row.storeId,
      status: row.status as SessionStatus,
      itemHash: row.itemHash,
      items: row.items.map(
        (i) =>
          new SessionItem(
            i.id,
            i.productId,
            i.barcode,
            i.quantity,
            i.pointsValue,
          ),
      ),
      createdAt: row.createdAt,
      updatedAt: row.updatedAt,
    });
  }

  /** Domain entity → Prisma session upsert payload (items handled separately). */
  static toPersistence(
    session: ShoppingSession,
  ): Prisma.ShoppingSessionUncheckedCreateInput {
    return {
      id: session.id,
      userId: session.userId,
      storeId: session.storeId,
      status: session.status,
      itemHash: session.itemHash,
      createdAt: session.createdAt,
      updatedAt: session.updatedAt,
    };
  }

  /** Domain entity → SessionItem rows for (re)insertion under the session. */
  static itemsToPersistence(
    session: ShoppingSession,
  ): Prisma.SessionItemUncheckedCreateInput[] {
    return session.items.map((i) => ({
      id: i.id,
      sessionId: session.id,
      productId: i.productId,
      barcode: i.barcode,
      quantity: i.quantity,
      pointsValue: i.pointsValue,
    }));
  }

  // ── Cache (de)serialization ──────────────────────────────────────────────────

  static toSnapshot(session: ShoppingSession): SessionSnapshot {
    return {
      id: session.id,
      userId: session.userId,
      storeId: session.storeId,
      status: session.status,
      itemHash: session.itemHash,
      createdAt: session.createdAt.toISOString(),
      updatedAt: session.updatedAt.toISOString(),
      items: session.items.map((i) => ({
        id: i.id,
        productId: i.productId,
        barcode: i.barcode,
        quantity: i.quantity,
        pointsValue: i.pointsValue,
      })),
    };
  }

  static fromSnapshot(snap: SessionSnapshot): ShoppingSession {
    return ShoppingSession.rehydrate({
      id: snap.id,
      userId: snap.userId,
      storeId: snap.storeId,
      status: snap.status,
      itemHash: snap.itemHash,
      items: snap.items.map(
        (i) =>
          new SessionItem(
            i.id,
            i.productId,
            i.barcode,
            i.quantity,
            i.pointsValue,
          ),
      ),
      createdAt: new Date(snap.createdAt),
      updatedAt: new Date(snap.updatedAt),
    });
  }
}
