// @smartcart/api/src/modules/checkout/infrastructure/crypto/jwt-qr.signer.ts
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import * as jwt from 'jsonwebtoken';
import { DomainError } from '../../domain/errors/checkout.errors';
import type {
  QrSignPayload,
  SignedQrToken,
} from '../../domain/factories/qr-ticket.factory';
import type {
  IQrSigner,
  QrTokenClaims,
} from '../../application/interfaces/qr-signer.interface';

/** Maps to QR_TOKEN_EXPIRED (422) — README §2.8 Workflow 1 error matrix. */
export class QrTokenExpiredError extends DomainError {
  readonly code = 'QR_TOKEN_EXPIRED';
  constructor() {
    super('The QR token has expired; generate a new one.');
  }
}

/** Maps to INVALID_QR_TOKEN (401) — forged / tampered signature. */
export class InvalidQrTokenError extends DomainError {
  readonly code = 'INVALID_QR_TOKEN';
  constructor() {
    super('The QR token signature is invalid.');
  }
}

/**
 * HS256 QR signer (README §2.3 QR rules, §2.5 A08).
 * - 10-minute expiry is set ONCE here (`expiresIn: '10m'`); the factory reads it
 *   back from the decoded `exp` claim, so there is a single source of truth.
 * - 10-second clock-skew tolerance on verify.
 * - Secret length (≥ 32) is enforced at startup by `env.validation.ts`.
 */
@Injectable()
export class JwtQrSigner implements IQrSigner {
  private readonly secret: string;

  constructor(config: ConfigService) {
    this.secret = config.getOrThrow<string>('QR_SIGNING_SECRET');
  }

  sign(payload: QrSignPayload): SignedQrToken {
    const token = jwt.sign(payload, this.secret, {
      algorithm: 'HS256',
      expiresIn: '10m',
    });
    // Read expiry back from the token — never compute it independently.
    const decoded = jwt.decode(token) as { exp: number };
    return { token, expiresAt: new Date(decoded.exp * 1000) };
  }

  verify(token: string): QrTokenClaims {
    try {
      return jwt.verify(token, this.secret, {
        algorithms: ['HS256'],
        clockTolerance: 10, // seconds — tolerate minor clock drift across instances
      }) as QrTokenClaims;
    } catch (err) {
      if (err instanceof jwt.TokenExpiredError) {
        throw new QrTokenExpiredError();
      }
      // JsonWebTokenError, NotBeforeError, malformed, wrong signature, etc.
      throw new InvalidQrTokenError();
    }
  }
}
