// @smartcart/api/src/common/health/health.controller.ts
import { Controller, Get } from '@nestjs/common';
import {
  HealthCheck,
  HealthCheckError,
  HealthCheckService,
  type HealthIndicatorResult,
} from '@nestjs/terminus';
import { PrismaService } from '../prisma/prisma.service';
import { RedisService } from '../redis/redis.service';

/**
 * Three-tier health probes (README §2.6 Health Checks, §2.7 availability).
 * - /health           — full dependency check (load balancer).
 * - /health/liveness   — lightweight process check (platform restart probe).
 * - /health/readiness  — DB + Redis connectivity (traffic routing).
 */
@Controller('health')
export class HealthController {
  constructor(
    private readonly health: HealthCheckService,
    private readonly prisma: PrismaService,
    private readonly redis: RedisService,
  ) {}

  @Get()
  @HealthCheck()
  check() {
    return this.health.check([() => this.pingDb(), () => this.pingRedis()]);
  }

  @Get('liveness')
  liveness(): { status: string } {
    return { status: 'ok' };
  }

  @Get('readiness')
  @HealthCheck()
  readiness() {
    return this.health.check([() => this.pingDb(), () => this.pingRedis()]);
  }

  private async pingDb(): Promise<HealthIndicatorResult> {
    try {
      await this.prisma.$queryRaw`SELECT 1`;
      return { database: { status: 'up' } };
    } catch (err) {
      throw new HealthCheckError('Database unavailable', {
        database: { status: 'down', message: String(err) },
      });
    }
  }

  private async pingRedis(): Promise<HealthIndicatorResult> {
    try {
      const pong = await this.redis.ping();
      if (pong !== 'PONG') throw new Error(`unexpected reply: ${pong}`);
      return { redis: { status: 'up' } };
    } catch (err) {
      throw new HealthCheckError('Redis unavailable', {
        redis: { status: 'down', message: String(err) },
      });
    }
  }
}
