// @smartcart/api/src/modules/analytics/application/services/analytics.service.ts
import { Injectable } from '@nestjs/common';
import type {
  ProductInsightsDTO,
  SegmentDistributionDTO,
  StoreOverviewDTO,
} from '@smartcart/shared-types';
import { PrismaService } from '../../../../common/prisma/prisma.service';
import { RedisService } from '../../../../common/redis/redis.service';

const SEGMENTS_TTL = 3600; // cache B2B distribution for 1h (README §2.3 step 6)
const K_ANONYMITY_MIN = 50; // segments < 50 users collapse into "other"

const storeKey = (storeId: string) => `analytics:store:${storeId}:segments`;
const GLOBAL_KEY = 'analytics:global:segment-distribution';

/**
 * B2B analytics service (README §2.3 step 6, §2.8 Workflow 2). All output is
 * aggregated/anonymized: any segment with < 50 users is merged into "other"
 * (k-anonymity) and no individual `userId` ever leaves the service (§2.5 A04/PII).
 */
@Injectable()
export class AnalyticsService {
  constructor(
    private readonly prisma: PrismaService,
    private readonly redis: RedisService,
  ) {}

  async getSegmentDistribution(
    storeId?: string,
  ): Promise<SegmentDistributionDTO> {
    const cacheKey = storeId ? storeKey(storeId) : GLOBAL_KEY;
    const cached = await this.redis.get(cacheKey);
    if (cached) return JSON.parse(cached) as SegmentDistributionDTO;

    // Scope to a store by the set of users who shopped there (ConsumerSegment
    // holds no storeId — it is one aggregated row per user, README §2.4).
    let where: { userId?: { in: string[] } } = {};
    if (storeId) {
      const shoppers = await this.prisma.shoppingSession.findMany({
        where: { storeId },
        select: { userId: true },
        distinct: ['userId'],
      });
      where = { userId: { in: shoppers.map((s) => s.userId) } };
    }

    const groups = await this.prisma.consumerSegment.groupBy({
      by: ['segmentName'],
      where,
      _count: { _all: true },
    });

    // k-anonymity merge.
    let other = 0;
    const named: { segmentName: string; count: number }[] = [];
    for (const g of groups) {
      const count = g._count._all;
      if (count < K_ANONYMITY_MIN) other += count;
      else named.push({ segmentName: g.segmentName, count });
    }
    if (other > 0) named.push({ segmentName: 'other', count: other });

    const total = named.reduce((sum, s) => sum + s.count, 0);
    const result: SegmentDistributionDTO = {
      segments: named.map((s) => ({
        segmentName: s.segmentName,
        count: s.count,
        percentage: total === 0 ? 0 : Math.round((s.count / total) * 10000) / 100,
      })),
    };

    await this.redis.set(cacheKey, JSON.stringify(result), 'EX', SEGMENTS_TTL);
    return result;
  }

  /**
   * README §2.4 endpoint; response fields are TBD in the spec, so this returns
   * the minimal documented shape rather than inventing metrics (see
   * `ProductInsightsDTO` in @smartcart/shared-types).
   */
  async getProductInsights(productId: string): Promise<ProductInsightsDTO> {
    return { productId };
  }

  /** README §2.4 endpoint; fields TBD per the spec (see `StoreOverviewDTO`). */
  async getStoreOverview(storeId: string): Promise<StoreOverviewDTO> {
    return { storeId };
  }
}
