// @smartcart/api/src/modules/checkout/infrastructure/repositories/prisma-session.repository.ts
import { Injectable } from '@nestjs/common';
import type { Prisma } from '@prisma/client';
import { PrismaService } from '../../../../common/prisma/prisma.service';
import { RedisService } from '../../../../common/redis/redis.service';
import type { ShoppingSession } from '../../domain/entities/shopping-session.entity';
import type {
  ISessionRepository,
  TransactionContext,
} from '../../application/interfaces/session-repository.interface';
import { SessionMapper, SessionSnapshot } from '../mappers/session.mapper';

/** Session cache TTL — matches the 2-hour max session lifetime (README §2.7). */
const SESSION_TTL_SECONDS = 7200;
const cacheKey = (id: string) => `session:${id}`;

/**
 * Prisma + Redis session repository (README §2.2 Rule 3, §2.7 stateless).
 * Read-through cache on `findById`; writes invalidate the cache (safe under
 * rollback — a cache miss simply re-reads the authoritative DB row).
 */
@Injectable()
export class PrismaSessionRepository implements ISessionRepository {
  constructor(
    private readonly prisma: PrismaService,
    private readonly redis: RedisService,
  ) {}

  /** Narrow the opaque tx handle back to a Prisma client (or fall back). */
  private db(tx?: TransactionContext): Prisma.TransactionClient | PrismaService {
    return (tx as Prisma.TransactionClient) ?? this.prisma;
  }

  async findById(
    id: string,
    tx?: TransactionContext,
  ): Promise<ShoppingSession | null> {
    // Skip the cache while inside a transaction — read the live row.
    if (!tx) {
      const cached = await this.redis.get(cacheKey(id));
      if (cached) {
        return SessionMapper.fromSnapshot(JSON.parse(cached) as SessionSnapshot);
      }
    }

    const row = await this.db(tx).shoppingSession.findUnique({
      where: { id },
      include: { items: true },
    });
    if (!row) return null;

    const session = SessionMapper.toDomain(row);
    if (!tx) {
      await this.redis.set(
        cacheKey(id),
        JSON.stringify(SessionMapper.toSnapshot(session)),
        'EX',
        SESSION_TTL_SECONDS,
      );
    }
    return session;
  }

  async findActiveByUser(userId: string): Promise<ShoppingSession | null> {
    const row = await this.prisma.shoppingSession.findFirst({
      where: { userId, status: 'ACTIVE' },
      include: { items: true },
    });
    return row ? SessionMapper.toDomain(row) : null;
  }

  async findActiveOlderThan(threshold: Date): Promise<ShoppingSession[]> {
    const rows = await this.prisma.shoppingSession.findMany({
      where: { status: 'ACTIVE', createdAt: { lt: threshold } },
      include: { items: true },
    });
    return rows.map((r) => SessionMapper.toDomain(r));
  }

  async save(session: ShoppingSession, tx?: TransactionContext): Promise<void> {
    const db = this.db(tx);
    const data = SessionMapper.toPersistence(session);

    await db.shoppingSession.upsert({
      where: { id: session.id },
      create: data,
      update: {
        status: data.status,
        itemHash: data.itemHash,
        updatedAt: data.updatedAt,
      },
    });

    // Replace the line items (cart is small; add/remove are infrequent).
    await db.sessionItem.deleteMany({ where: { sessionId: session.id } });
    const items = SessionMapper.itemsToPersistence(session);
    if (items.length > 0) {
      await db.sessionItem.createMany({ data: items });
    }

    // Invalidate cache — safe regardless of commit/rollback (re-reads on miss).
    await this.redis.del(cacheKey(session.id));
  }
}
