// @smartcart/api/src/modules/checkout/infrastructure/repositories/prisma-points.repository.ts
import { Injectable } from '@nestjs/common';
import type { Prisma, PointsReason as PrismaPointsReason } from '@prisma/client';
import { PrismaService } from '../../../../common/prisma/prisma.service';
import type {
  IPointsRepository,
  LedgerEntry,
} from '../../application/interfaces/points-repository.interface';
import type { TransactionContext } from '../../application/interfaces/unit-of-work.interface';

/**
 * Append-only points ledger (README §2.4 PointsTransaction, §2.5 A04/A08).
 * Crediting = inserting an immutable row. Balance is DERIVED via `SUM(delta)`;
 * there is no UPDATE path on any balance, so records cannot be silently rewritten.
 */
@Injectable()
export class PrismaPointsRepository implements IPointsRepository {
  constructor(private readonly prisma: PrismaService) {}

  private db(tx?: TransactionContext): Prisma.TransactionClient | PrismaService {
    return (tx as Prisma.TransactionClient) ?? this.prisma;
  }

  async creditPoints(
    entry: LedgerEntry,
    tx?: TransactionContext,
  ): Promise<void> {
    await this.db(tx).pointsTransaction.create({
      data: {
        userId: entry.userId,
        delta: entry.delta,
        reason: entry.reason as PrismaPointsReason,
        sessionId: entry.sessionId ?? null,
      },
    });
  }

  async getBalance(userId: string, tx?: TransactionContext): Promise<number> {
    const result = await this.db(tx).pointsTransaction.aggregate({
      where: { userId },
      _sum: { delta: true },
    });
    return result._sum.delta ?? 0;
  }
}
