// @smartcart/api/src/modules/checkout/domain/state-machine/session-state-machine.ts
import type { SessionStatus } from '@smartcart/shared-types';
import { InvalidTransitionError } from '../errors/checkout.errors';

/**
 * Finite-state machine for the shopping-session lifecycle (README §2.3 / §2.4).
 * Pure domain — encodes ONLY the legal state graph. Contextual guards
 * (items > 0, age > 2h) are enforced by the entity / cron, per the README table.
 *
 *   ACTIVE ──addItem──▶ ACTIVE
 *   ACTIVE ──requestCheckout──▶ PENDING_CHECKOUT
 *   ACTIVE / PENDING_CHECKOUT ──expire──▶ EXPIRED
 *   PENDING_CHECKOUT ──completeValidation──▶ COMPLETED
 *   PENDING_CHECKOUT ──markValidationFailed──▶ VALIDATION_FAILED
 *   COMPLETED ──expire──▶ COMPLETED        (idempotent, no transition)
 */
export enum SessionEvent {
  ADD_ITEM = 'addItem',
  REQUEST_CHECKOUT = 'requestCheckout',
  COMPLETE_VALIDATION = 'completeValidation',
  MARK_VALIDATION_FAILED = 'markValidationFailed',
  EXPIRE = 'expire',
}

type TransitionTable = {
  [S in SessionStatus]?: Partial<Record<SessionEvent, SessionStatus>>;
};

const TRANSITIONS: TransitionTable = {
  ACTIVE: {
    [SessionEvent.ADD_ITEM]: 'ACTIVE',
    [SessionEvent.REQUEST_CHECKOUT]: 'PENDING_CHECKOUT',
    [SessionEvent.EXPIRE]: 'EXPIRED',
  },
  PENDING_CHECKOUT: {
    [SessionEvent.COMPLETE_VALIDATION]: 'COMPLETED',
    [SessionEvent.MARK_VALIDATION_FAILED]: 'VALIDATION_FAILED',
    [SessionEvent.EXPIRE]: 'EXPIRED',
  },
  COMPLETED: {
    // Idempotent: expiring a completed session is a no-op, not an error.
    [SessionEvent.EXPIRE]: 'COMPLETED',
  },
};

export class SessionStateMachine {
  /** True if `event` is legal from `from`. */
  static guardTransition(from: SessionStatus, event: SessionEvent): boolean {
    return TRANSITIONS[from]?.[event] !== undefined;
  }

  /** Resolve the next status, or throw `InvalidTransitionError` if illegal. */
  static transition(from: SessionStatus, event: SessionEvent): SessionStatus {
    const next = TRANSITIONS[from]?.[event];
    if (next === undefined) {
      throw new InvalidTransitionError(from, event);
    }
    return next;
  }
}
